### PR TITLE
Issue #508: Implement tracking protection in SystemEngine

### PR DESCRIPTION
**Pull request by csadilek**
The meat of the tracking protection logic in Focus is in the URL matching and black/whitelisting logic. I was considering making this an application-level concern, but I think, by default, our components should support tracking protection i.e. if `enableTrackingProtection` is called it should just work without the App having to provide the URL matching functionality.

So, I ported all this logic to Kotlin, refactored it and wrote tests, then added `enable|disableTrackingProtection` to our `SystemEngine`.

### CODE REVIEW

**Comment by csadilek** [comment]
@pocmo OK, this is updated now to move the matcher functionality to the system engine.

File: components/browser/engine-system/build.gradle [hunk 1]
```diff
@@ -32,6 +32,7 @@ dependencies {
     implementation project(':support-utils')
 
     implementation "org.jetbrains.kotlin:kotlin-stdlib:${rootProject.ext.dependencies['kotlin']}"
+    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:${rootProject.ext.dependencies['coroutines']}"
```
Comment (pocmo) [inline components/browser/engine-system/build.gradle]: I need to update this before #636 lands (or the other way around).

Comment (csadilek) [reply components/browser/engine-system/build.gradle]: Ah OK, I will do it. I will merge #636 and update this.

**Comment by csadilek** [comment]
OK, updated!

**Pull request merged**